otelconf/v0.3.0/metric_test.go

```diff
@@ -1388,3 +1394,188 @@ func TestPrometheusIPv6(t *testing.T) {
 		})
 	}
 }
+
+func Test_otlpGRPCMetricExporter(t *testing.T) {
+	type args struct {
+		ctx        context.Context
+		otlpConfig *OTLPMetric
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr string
+	}{
+		{
+			name: "no TLS config",
+			args: args{
+				ctx: context.Background(),
+				otlpConfig: &OTLPMetric{
+					Protocol:    ptr("grpc"),
+					Compression: ptr("gzip"),
+					Timeout:     ptr(5000),
+					Insecure:    ptr(true),
+					Headers: []NameStringValuePair{
+						{Name: "test", Value: ptr("test1")},
+					},
+				},
+			},
+		},
+		{
+			name: "with TLS config",
+			args: args{
+				ctx: context.Background(),
+				otlpConfig: &OTLPMetric{
+					Protocol:    ptr("grpc"),
+					Compression: ptr("gzip"),
+					Timeout:     ptr(5000),
+					Certificate: ptr("testdata/server-certs/server.crt"),
+					Headers: []NameStringValuePair{
+						{Name: "test", Value: ptr("test1")},
+					},
+				},
+			},
+		},
+		{
+			name: "with TLS config and client key",
+			args: args{
+				ctx: context.Background(),
+				otlpConfig: &OTLPMetric{
+					Protocol:          ptr("grpc"),
+					Compression:       ptr("gzip"),
+					Timeout:           ptr(5000),
+					Certificate:       ptr("testdata/server-certs/server.crt"),
+					ClientKey:         ptr("testdata/client-certs/client.key"),
+					ClientCertificate: ptr("testdata/client-certs/client.crt"),
+					Headers: []NameStringValuePair{
+						{Name: "test", Value: ptr("test1")},
+					},
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			port, err := findRandomPort()
+			require.NoError(t, err)
+
+			tt.args.otlpConfig.Endpoint = ptr(fmt.Sprintf("localhost:%d", port))
+
+			tlsMode := ""
+			if tt.args.otlpConfig.Insecure == nil || !*tt.args.otlpConfig.Insecure {
+				tlsMode = "TLS"
+			}
+			if tt.args.otlpConfig.ClientCertificate != nil && *tt.args.otlpConfig.ClientCertificate != "" {
+				tlsMode = "mTLS"
+			}
+			col, err := newGRPCMetricCollector(*tt.args.otlpConfig.Endpoint, tlsMode)
+			require.NoError(t, err)
+
+			exporter, err := otlpGRPCMetricExporter(tt.args.ctx, tt.args.otlpConfig)
+			if tt.wantErr != "" {
+				require.Error(t, err)
+				require.Equal(t, tt.wantErr, err.Error())
+			} else {
+				require.NoError(t, err)
+			}
+
+			res, err := resource.New(context.Background())
+			require.NoError(t, err)
+
+			require.EventuallyWithT(t, func(collect *assert.CollectT) {
+				require.NoError(collect, exporter.Export(context.Background(), &metricdata.ResourceMetrics{
+					Resource: res,
+					ScopeMetrics: []metricdata.ScopeMetrics{
+						{
+							Metrics: []metricdata.Metrics{
+								{
+									Name: "test-metric",
+									Data: metricdata.Gauge[int64]{
+										DataPoints: []metricdata.DataPoint[int64]{
+											{
+												Value: 1,
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				}))
+			}, 10*time.Second, 1*time.Second)
```

Comment:
Cannot we call it synchronously?

Reply:
theoretically there could be some timing issues, as the server is started in a separate routine. But I will try again with calling this synchronously (need to run it a couple of times to make sure we don't introduce some flakiness)

Reply:
after removing the `require.Eventually` one of the test runs ended up failing, so I have re-added it again. It seems like it is indeed needed, due to the reason outlined in the comment above